mitiq/pec/pec.py

```diff
@@ -0,0 +1,115 @@
+    circuit: QPROGRAM,
+    executor: Callable[[QPROGRAM], float],
+    deco_dict: DecoType,
+    num_to_average: int = 1,
+    num_samples: Optional[int] = None,
+) -> float:
+    """Evaluates the expectation value associated to the input circuit
+    using probabilistic error cancellation (PEC) [Temme2017]_.
+
+    This function implements PEC by:
+
+    1. Sampling different implementable circuits from the quasi-probability
+       representation of the input circuit;
+    2. Evaluating the noisy expectation values associated to the sampled
+       circuits (through the "executor" function provided by the user);
+    3. Estimating the ideal expectation value from a suitable linear
+       combination of the noisy ones.
+
+    Args:
+        circuit = The input circuit to execute with error-mitigation.
+        executor = A function which executes a circuit and returns an
+            expectation value.
+        deco_dict = The decomposition dictionary containing the quasi-
+            probability representation of the ideal operations (those
+            which are part of the input circuit).
+        num_to_average: The number times the expectation value of each sampled
+            circuit is evaluated, then averaged. Setting this number to values
+            larger than 1 (default), reduces the statistical error associated
+            to each noisy expectation value.
+        num_samples: The number of noisy circuits to be sampled for PEC.
+            If equal to None, it is deduced from the amount of "negativity"
+            of the quasi-probability representation of the input circuit.
+            Note: the latter feature is not yet implemented and num_samples
+            is just set to 1000 if not specified.
+
+    Returns:
+        The PEC estimate of the ideal expectation value associated
+        to the input circuit.
+
+    .. [Temme2017] : Kristan Temme, Sergey Bravyi, Jay M. Gambetta,
+        "Error mitigation for short-depth quantum circuits,"
+        *Phys. Rev. Lett.* **119**, 180509 (2017),
+        (https://arxiv.org/abs/1612.02058).
+
+    .. [Endo2018] : Suguru Endo, Simon C. Benjamin, Ying Li,
+        "Practical Quantum Error Mitigation for Near-Future Applications"
+        *Phys. Rev. **X 8**, 031027 (2018),
+        (https://arxiv.org/abs/1712.09271).
+
+    .. [Takagi2020] : Ryuji Takagi,
+        "Optimal resource cost for error mitigation,"
+        (https://arxiv.org/abs/2006.12509).
+    """
+
+    # TODO: maybe deduce num_to_sample from the decomposition "negativity"
+    if not num_samples:
+        num_samples = 1000
+
+    sampled_circuits = []
+    signs = []
+    norms = []
+
+    for _ in range(num_samples):
+        sampled_circuit, sign, norm = sample_circuit(circuit, deco_dict)
+        sampled_circuits.append(sampled_circuit)
+        signs.append(sign)
+        norms.append(norm)
+
+    # The norm of the quasi-distribution should be independent of sample
+    assert np.all(norms == norms[0])
```

Comment:
I see from `sample_circuit` below that it is indeed useful to return the norm with the current setup, since it's different for different circuits but here the same circuit is passed in every time. So these comments can be ignored for now.